validator/src/main/java/com/amazon/aoc/validators/TraceValidator.java

```diff
@@ -19,97 +19,99 @@
-          if (expectedTraceList.size() != traceList.size()) {
-            throw new BaseException(ExceptionCode.TRACE_LIST_NOT_MATCHED);
-          }
-
-          for (int i = 0; i != expectedTraceList.size(); ++i) {
-            Trace trace = traceList.get(i);
-            compareTwoTraces(expectedTraceList.get(i), trace);
-          }
-        });
-  }
-
-  private void compareTwoTraces(Trace trace1, Trace trace2) throws BaseException {
-    // check trace id
-    if (!trace1.getId().equals(trace2.getId())) {
-      throw new BaseException(ExceptionCode.TRACE_ID_NOT_MATCHED);
+    @Override
+    public void init(Context context, ICaller caller, FileConfig expectedTrace) throws Exception {
+        this.xrayService = new XRayService(context.getRegion());
+        this.caller = caller;
+        this.expectedTrace = expectedTrace;
     }
 
-    /*
-    if (trace1.getSegments().size() != trace2.getSegments().size()) {
-      throw new BaseException(ExceptionCode.TRACE_SPAN_LIST_NOT_MATCHED);
+    @Override
+    public void validate() throws Exception {
+        // get stored trace
+        Map<String, Object> storedTrace = this.getStoredTrace();
+
+        // create trace id list to retrieve trace from x-ray service
+        String traceId = (String) storedTrace.get("trace_id");
+        List<String> traceIdList = Collections.singletonList(traceId);
+
+        // get retrieved trace from x-ray service
+        Map<String, Object> retrievedTrace = this.getRetrievedTrace(traceIdList);
+
+        // validation of trace id
+        if (!storedTrace.get("trace_id").equals(retrievedTrace.get("trace_id"))) {
+            throw new BaseException(ExceptionCode.TRACE_ID_NOT_MATCHED);
+        }
+
+        // data model validation of other fields of segment document
+        for (Map.Entry<String, Object> entry : storedTrace.entrySet()) {
+            if (!entry.getValue().equals(retrievedTrace.get(entry.getKey()))) {
+                log.info("mis matched data model field list");
+                log.info("value of stored trace map: {}", entry.getValue());
+                log.info("value of retrieved map: {}",retrievedTrace.get(entry.getKey()));
+                log.info("==========================================");
+                throw new BaseException(ExceptionCode.DATA_MODEL_NOT_MATCHED);
+            }
+        }
+    }
+
+    // this method will hit get trace from x-ray service and get retrieved trace
+    private Map<String, Object> getRetrievedTrace(List<String> traceIdList) throws Exception {
+        Map<String, Object> flattenedJsonMapForRetrievedTrace = null;
+
+        // retrieve trace from x-ray service
+        TimeUnit.SECONDS.sleep(15);
+        List<Trace> retrieveTraceList = xrayService.listTraceByIds(traceIdList);
+
+        try {
+            // flattened JSON object to a map
+            flattenedJsonMapForRetrievedTrace = JsonFlattener.flattenAsMap(retrieveTraceList.get(0).getSegments().get(0).getDocument());
```

Comment:
it will be better if we can do null point check here

Reply:
good suggestion!